### PR TITLE
Add buffer pool watermark support

### PR DESCRIPTION
**Pull request by wendani**
1. Write fields to FLEX_COUNTER_GROUP_TABLE at the construction of a BufferOrch object:
    
    "POLL_INTERVAL"
    "BUFFER_POLL_PLUGIN_LIST"
    "STATS_MODE"

2. Update buffer pool name to oid mapping in COUNTERS_DB upon the set and del of its oid

3. Push buffer pool watermark COUNTER_ID_LIST to FLEX_COUNTER_TABLE

4. Implement user clear logic to buffer pool watermark

5. Add periodic clear to buffer pool watermark

6. Add lua script for watermark_bufferpool

**What I did**

**Why I did it**

**How I verified it**
On brcm dut
**Details if related**
Depend on:
https://github.com/Azure/sonic-swss-common/pull/272 Add schema for buffer pool watermark
https://github.com/Azure/sonic-buildimage/pull/2818 Feed device info to orchagent process

Alongside:
https://github.com/Azure/sonic-sairedis/pull/451 Add buffer pool stat to flex counter architecture

Downstream changes:
https://github.com/Azure/sonic-utilities/pull/521 Add buffer pool watermark support in watermarkstat and counterpoll

syncd rpc support:
https://github.com/opencomputeproject/SAI/pull/949

### CODE REVIEW

**Comment by wendani** [comment]
@mykolaf for review

File: orchagent/bufferorch.cpp [hunk 1]
```diff
@@ -82,6 +97,38 @@ void BufferOrch::initBufferReadyList(Table& table)
...
+        fvTuples.emplace_back(BUFFER_POOL_PLUGIN_FIELD, bufferPoolWmSha);
+        fvTuples.emplace_back(POLL_INTERVAL_FIELD, BUFFER_POOL_WATERMARK_FLEX_STAT_COUNTER_POLL_MSECS);
+
+        string statsMode = STATS_MODE_READ_AND_CLEAR;
+        // Some platforms do not support buffer pool watermark clear operation
+        // Need the SAI API support to query the capability
+        // Before this capability is in place, we feed the platform info into the orchagent
+        // daemon and check against it
+        char *device = getenv("device");
+        if (device && (strstr(device, "7050") || strstr(device, "6000")))
```
Comment (qiluo-msft) [inline orchagent/bufferorch.cpp]: >7050 [](start = 39, length = 4)

Please be more specific about device strings.

File: orchagent/bufferorch.h [hunk 2]
```diff
@@ -50,6 +58,15 @@ class BufferOrch : public Orch
     buffer_table_handler_map m_bufferHandlerMap;
     std::unordered_map<std::string, bool> m_ready_list;
     std::unordered_map<std::string, std::vector<std::string>> m_port_ready_list_ref;
+
+    unique_ptr<DBConnector> m_flexCounterDb = nullptr;
```
Comment (qiluo-msft) [inline orchagent/bufferorch.h]: > = nullptr [](start = 43, length = 10)

No need to initialize.

File: orchagent/bufferorch.cpp [hunk 3]
```diff
@@ -82,6 +97,38 @@ void BufferOrch::initBufferReadyList(Table& table)
...
+        {
+            statsMode = STATS_MODE_READ;
+        }
+        fvTuples.emplace_back(STATS_MODE_FIELD, statsMode);
+        m_flexCounterGroupTable->set(BUFFER_POOL_WATERMARK_STAT_COUNTER_FLEX_COUNTER_GROUP, fvTuples);
+    }
+    catch (...)
```
Comment (qiluo-msft) [inline orchagent/bufferorch.cpp]: >catch (...) [](start = 4, length = 11)

Dangerous to catch all.

**Comment by lguohan** [comment]
retest this please

File: orchagent/bufferorch.cpp [hunk 4]
```diff
@@ -105,6 +152,51 @@ bool BufferOrch::isPortReady(const std::string& port_name) const
...
+        return;
+    }
+
+    // Detokenize the SAI watermark stats to a string, separated by comma
+    string statList;
+    for (const auto &it : bufferPoolWatermarkStatIds)
+    {
+        statList += (sai_serialize_buffer_pool_stat(it) + list_item_delimiter);
+    }
+    if (!statList.empty())
```
Comment (mykolaf) [inline orchagent/bufferorch.cpp]: This is redundant. The FlexCounter will consume a COUNTER_ID_LIST with a trailing comma.

Comment (qiluo-msft) [reply orchagent/bufferorch.cpp]: Even this is true, it is still recommended to remove it

---
In reply to: [278052238](https://github.com/Azure/sonic-swss/pull/853#discussion_r278052238) [](ancestors = 278052238)

**Comment by wendani** [comment]
retest this please

File: orchagent/bufferorch.cpp [hunk 5]
```diff
@@ -82,6 +97,32 @@ void BufferOrch::initBufferReadyList(Table& table)
...
+        // TODO (work in progress):
+        // Some platforms do not support buffer pool watermark clear operation on a particular pool
+        // Invoke the SAI clear_stats API per pool to query the capability from the API call return status
+        fvTuples.emplace_back(STATS_MODE_FIELD, STATS_MODE_READ_AND_CLEAR);
+
+        m_flexCounterGroupTable->set(BUFFER_POOL_WATERMARK_STAT_COUNTER_FLEX_COUNTER_GROUP, fvTuples);
+    }
+    catch (const exception &e)
+    {
+        SWSS_LOG_WARN("Buffer pool watermark lua script and/or flex counter group not set successfully. Runtime error: %s", e.what());
```
Comment (qiluo-msft) [inline orchagent/bufferorch.cpp]: >SWSS_LOG_WARN [](start = 8, length = 13)

You said it is a runtime error, why using WARN?

File: orchagent/bufferorch.cpp (hunk 3 above)
Comment (qiluo-msft) [reply orchagent/bufferorch.cpp]: Still too general, could you be more specific?

---
In reply to: [277914027](https://github.com/Azure/sonic-swss/pull/853#discussion_r277914027) [](ancestors = 277914027)

**Comment by wendani** [comment]
retest this please

**Comment by wendani** [comment]
retest this please

**Comment by wendani** [comment]
retest this please

**Comment by wendani** [comment]
retest this please

**Comment by wendani** [comment]
retest this please

**Pull request merged**